File: java/graphs/medium/NumberOfIslands.java
class Solution {
    public int numIslands(char[][] grid) {
        int islandNum = 0;
        boolean[][] isVisited = new boolean[grid.length][grid[0].length];
        for(int i = 0; i < grid.length; i++) {
            for(int j = 0; j < grid[0].length; j++) {
                if(grid[i][j] == '1' && !isVisited[i][j]) {
                    islandNum += 1;
                    dfs(grid, i, j, isVisited);
                }
            }
        }
        
        return islandNum;
    }
    
    public void dfs(char[][] grid, int row, int col, boolean[][]isVisited) {
        if(row < 0 || row > grid.length - 1 || col < 0 || col > grid[0].length - 1 || grid[row][col] == '0' || isVisited[row][col] == true) {return;}
        
        grid[row][col] = '0';
        isVisited[row][col] = true;
        dfs(grid, row, col + 1, isVisited);
        dfs(grid, row + 1, col, isVisited);
        dfs(grid, row, col - 1, isVisited);
        dfs(grid, row - 1, col, isVisited);
    }
}

File: Python/Graphs/Medium/numberOfIslands.py
class Solution(object):
    def islands(self, grid):
        # init islands, visited set, rows/cols of grid
        islands = 0
        visited = set()
        rows, cols = len(grid), len(grid[0])

        # case to see if grid is empty 
        if not grid:
            return 0 

        # dfs function 
        def dfs(r, c):
            # conditions that will return nothing 
            if (r not in range(rows) or
                c not in range(cols) or 
                grid[r][c] == "0" or 
                (r,c) in visited):
                return 

            visited.add((r,c))
            # check all 4 directions 
            directions = [[1,0],[0,1],[-1,0],[0,-1]]
            for dr, dc in directions:
                dfs(r + dr, c + dc)

        # iterate through entire grid and update number of islands 
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1" and (r,c) not in visited:
                    islands += 1
                    dfs(r,c)
        return islands